lib/Dialect/FIRRTL/Transforms/LowerBundleVectorTypes.cpp

```diff
@@ -0,0 +1,958 @@
+          auto elt = bundle.getElement(i);
+          // Construct the suffix to pass down.
+          tmpSuffix.resize(0);
+          tmpSuffix.push_back('_');
+          tmpSuffix.append(elt.name.getValue());
+          fields.emplace_back(elt.type, i, bundle.getFieldID(i),
+                              elt.type.getMaxFieldID()+1, tmpSuffix, elt.isFlip);
+        }
+        return true;
+      })
+      .Case<FVectorType>([&](auto vector) {
+        // Increment the field ID to point to the first element.
+        auto width = vector.getElementType().getMaxFieldID() + 1;
+        for (size_t i = 0, e = vector.getNumElements(); i != e; ++i) {
+          fields.emplace_back(vector.getElementType(), i, vector.getFieldID(i),
+                              width, "_" + std::to_string(i), false);
+        }
+        return true;
+      })
+      .Default([](auto op) { return false; });
+}
+
+static MemOp cloneMemWithNewType(ImplicitLocOpBuilder *b, MemOp op,
+                                 FIRRTLType type, StringRef suffix) {
+  SmallVector<Type, 8> ports;
+  SmallVector<Attribute, 8> portNames;
+  for (auto port : op.getPorts()) {
+    ports.push_back(MemOp::getTypeForPort(op.depth(), type, port.second));
+    portNames.push_back(port.first);
+  }
+  return b->create<MemOp>(ports, op.readLatency(), op.writeLatency(),
+                          op.depth(), op.ruw(), b->getArrayAttr(portNames),
+                          (op.name() + suffix).str(), op.annotations(),
+                          op.portAnnotations());
+}
+
+/// Look through and collect subfields leading to a subaccess.
+static SmallVector<Operation *> getSAWritePath(Operation *op) {
+  SmallVector<Operation *> retval;
+  auto defOp = op->getOperand(0).getDefiningOp();
+  while (defOp && isa<SubfieldOp, SubindexOp, SubaccessOp>(defOp)) {
+    retval.push_back(defOp);
+    defOp = defOp->getOperand(0).getDefiningOp();
+  }
+  // Trim to the subaccess
+  while (!retval.empty() && !isa<SubaccessOp>(retval.back()))
+    retval.pop_back();
+  return retval;
+}
+
+/// Copy annotations from \p annotations to \p loweredAttrs, except annotations
+/// with "target" key, that do not match the field suffix.
+static SmallVector<Attribute> filterAnnotations(MLIRContext *ctxt,
+                                                ArrayAttr annotations,
+                                                unsigned targetFieldID,
+                                                unsigned fieldWidth) {
+  SmallVector<Attribute> retval;
+  if (!annotations || annotations.empty())
+    return retval;
+
+  for (auto opAttr : annotations) {
+    if (auto subAnno = opAttr.dyn_cast<SubAnnotationAttr>()) {
+      if (targetFieldID <= subAnno.getMinFieldID() &&
+          targetFieldID + fieldWidth > subAnno.getMaxFieldID()) {
+        auto newMin = subAnno.getMinFieldID() - targetFieldID;
+        auto newMax = subAnno.getMaxFieldID() - targetFieldID;
```

Comment:
```suggestion
      if (targetMinFieldID >= subAnno.getMinFieldID() &&
          targetMaxFieldID <= subAnno.getMaxFieldID()) {
        auto newMin = subAnno.getMinFieldID() - targetMinFieldID;
        auto newMax = subAnno.getMaxFieldID() - targetMinFieldID;
```

Reply:
It was a bit more complex, it needed to look for the intersection of two ranges.